src/core/dom/image/lifecycle.ts

```diff
@@ -0,0 +1,217 @@
+/*!
+ * V4Fire Client Core
+ * https://github.com/V4Fire/Client
+ *
+ * Released under the MIT license
+ * https://github.com/V4Fire/Client/blob/master/LICENSE
+ */
+
+import ImageLoader from 'core/dom/image/image';
+
+import {
+
+	ImageNode,
+	ImagePlaceholderType,
+
+	IMG_IS_LOADED,
+	INIT_LOAD,
+	LOADING_STARTED,
+	ID
+
+} from 'core/dom/image';
+
+/**
+ * Helper class, provides an API to work with image lifecycle
+ */
+export default class Lifecycle {
+	/**
+	 * Parent class
+	 */
+	protected parent: ImageLoader;
+
+	/**
+	 * @param parent
+	 */
+	constructor(parent: ImageLoader) {
+		this.parent = parent;
+	}
+
+	/**
+	 * Initializes lifecycle of the specified el
+	 * @param el
+	 */
+	init(el: ImageNode): void {
+		const
+			previewShadowState = this.parent.getShadowStateByType(el, 'preview'),
+			mainShadowState = this.parent.getShadowStateByType(el, 'main');
+
+		if (mainShadowState?.mainOptions.stageClasses) {
+			this.parent.setClasses(el, mainShadowState, 'initial');
+		}
+
+		if (previewShadowState != null) {
+			this.initHelperImage(el, 'preview');
+		}
+
+		this.initMain(el);
+	}
+
+	/**
+	 * Initializes a main image
+	 * @param el
+	 */
+	protected initMain(el: ImageNode): void {
+		const mainShadowState = this.parent.getShadowStateByType(el, 'main');
+
+		if (mainShadowState == null) {
+			return;
+		}
+
+		if (mainShadowState.imgNode.complete === true && mainShadowState.imgNode[IMG_IS_LOADED] === true) {
+			this.onMainImageLoad(el);
+
+		} else {
+			const $a = mainShadowState.mainOptions.ctx?.unsafe.async;
+
+			if ($a) {
+				mainShadowState.loadPromise = $a.promise(mainShadowState.imgNode.init, {group: '[[v-image:main]]', label: el[ID]})
+					.then(this.onMainImageLoad.bind(this, el))
+					.catch(this.onMainImageLoadError.bind(this, el));
+
+			} else {
+				mainShadowState.loadPromise = mainShadowState.imgNode.init
+					.then(this.onMainImageLoad.bind(this, el))
+					.catch(this.onMainImageLoadError.bind(this, el));
+			}
+		}
+	}
+
+	/**
+	 * Initializes a preview image
+	 *
+	 * @param el
+	 * @param type
+	 */
+	protected initHelperImage(el: ImageNode, type: ImagePlaceholderType): void {
+		const
+			successCallback = this.trySetHelperImage.bind(this, el, type),
+			errorCallback = this.onHelperImageError.bind(this, el, type);
+
+		const
+			shadowState = this.parent.getShadowStateByType(el, type);
+
+		if (shadowState == null) {
+			return;
+		}
+
+		const
+			{mainOptions} = shadowState,
+			{imgNode} = shadowState;
+
+		if (imgNode[IMG_IS_LOADED] === true) {
+			/*
+			 * If an img is ready – set it to the element
+			 */
+			return successCallback();
+		}
+
+		if (imgNode[LOADING_STARTED] == null) {
+			/*
+			 * If loading was not started – this is a `broken` image that should be loaded lazy
```

Comment:
If loading hasn't started – this is the `broken` image that should be loaded lazily